### PR TITLE
Add operator and cast support for GeneratorParam.

### PR DESCRIPTION
**Pull request by zvookin**
Allows GeneratorParams to be used in expressions with arithmetic
types (integers, floats, etc.) and Expr. This was removed with the
recent changes to how immediates are handled in the IR and this restores
that functionality.

It is possible to use template metaprogramming to unify this with the operator overloads in IROperator.h, and it may be a good idea to do so. I've started working on an implementation of that and am not sure how long it is going to take. Hence I'd like to get consider this approach first and the more elaborate extensible one later. There is some risk of changing the language semantics slightly doing it this way so I am open to feedback here. The generalized code is good if we anticipate there being other types that need to interoperate with Expr and the arithmetic types in this way.

### CODE REVIEW

File: src/Generator.h [hunk 1]
```diff
@@ -328,6 +329,231 @@ template <typename T> class GeneratorParam : public Internal::GeneratorParamBase
     }
 };
 
+/** Addition between GeneratorParam and arithmetic/enum types.
+ * Returns arithmetic type. */
+// @{
+template <typename T1, typename T2, typename std::enable_if<std::is_arithmetic<T1>::value || std::is_enum<T1>::value>::type * = nullptr>
```
Comment (steven-johnson) [inline src/Generator.h]: Style nit: shorter line breaks would make this more readable IMHO

**Comment by steven-johnson** [comment]
We should makes sure there are no red flags with MSVC; otherwise this looks a not-unreasonable approach.

**Comment by abadams** [comment]
I'm confused about why the Expr versions exist separately when they have the same body as the arithmetic/enum versions. Why not just skip the enable_if entirely and say that GeneratorParam<T> combines with anything via conversion to T.

**Comment by kgnk** [comment]
Alternatively, add `|| std::is_same< T1, Halide::Expr >::value` to the `enable_if` condition and only have one set of operators.

**Comment by zvookin** [comment]
The suggestions on trying to get one set of operators is good. Let me give the goals so we can arrive at the right design:

1) GeneratorParam works as if it were a C type like int if used in a regular C++ expression not going into Halide IR.

2) GeneratorParam works as if it were an Expr when being used in Halide IR contexts.

3) 1 and 2 happen without any ambiguity errors. This is the reason we need overloads that take GeneratorParam<T> directly rather than allowing GeneratorParam<T> to be converted to T or Expr. If the conversion is there, it is very likely there will be two different overloads each one conversion away.

4) The reason I put in the enable_if tests is I felt like having GeneratorParam<T> export templates which work with arbitrary other types was bad. Ie. we can make it work with the basic types of the language and Expr but having it just compose with SomeTypeFromAnotherLibrary seemed problematic. Though now that I think about it, it seems like it might be ok. The other reason the enable_if might be needed is avoiding ambiguity with the operators in IROperator.h .

5) If I do keep the enable_if clause and or "is the other thing an Expr" into it, I'm not sure if I should use is_same or is_convertible to test that the other type is Expr (or Expr-like).

**Comment by abadams** [comment]
I'm not sure the code currently does point 2.

E.g., it looks like if you have a GeneratorParam<int> and add it to a
UInt(8) Expr, it'll go via operator+(Expr, int) and give you a UInt(8) Expr
again. Following point 2, and acting like an Expr with type Int(32) would
give you an Int(32) Expr.

I'm not sure which behavior is better though.

**Comment by zvookin** [comment]
I simplified this to use one set of operators that work between GeneratorParam<T> and any type that allows the relevant operator with T. (Getting min and max to work required a little ugliness with using std::min and std::max.) I also added test for all the operators between GeneratorParam<T> and Expr, int, and float.

File: src/Generator.h [hunk 2]
```diff
@@ -328,6 +330,137 @@ template <typename T> class GeneratorParam : public Internal::GeneratorParamBase
     }
 };
 
+/** Addition between GeneratorParam and arithmetic/enum types.
+ * Returns arithmetic type. */
```
Comment (abadams) [inline src/Generator.h]: These comments now seem incorrect. How about "GeneratorParam<T> acts like T for the purpose of all operator overloads".

**Comment by zvookin** [comment]
Comments improved.

**Comment by zvookin** [comment]
Up to date with master and now passing all tests.

**Pull request merged**